R/item_upload_files.R

```diff
@@ -120,3 +120,145 @@ multi_file_body <- function(files){
+	
+	gql <- httr::handle(url = pkg.env$graphql_url)
+	
+	session_id = create_multipart_upload_session(
+		f_path, mimetype, 
+		session_details(session = session)$username, gql)
+		
+	refresh_token_before_expired()
+	
+	f <- file(description = file, open = "rb")
+	
+	on.exit(close(f), add = TRUE)
+	
+	part_number <- 0
+	parts_header <- list()
+	
+	if(status)
+		pb <- txtProgressBar(0, f_chunks, 1, "-", width = 50, style = 3)
+	
+	while ( length(chunk <- readBin(con = f, what = "raw", 
+																	n = chunk_size_bytes)) > 0 ) {
+		part_number <- part_number + 1
+		
+		refresh_token_before_expired()
+		
+		presignedUrl <- get_presigned_url_for_chunk(f_path, session_id, part_number, gql)
+		
+		put_chunk <- httr::PUT(presignedUrl, data = chunk)
+		
+		if(!put_chunk$status_code == 200) stop(paste("Error uploading file. \n",
+																					 "status:", put_chunk$status_code,
+																					 "content:", rawToChar(put_chunk$content)))
+		parts_header[[part_number]] <- list(ETag = put_chunk$headers$ETag,
+																				PartNumber = part_number)
+		if(status)
+			setTxtProgressBar(pb, part_number)
+	}
+	
+	if(status)
+		close(pb)
+	
+	return(invisible(complete_multipart_upload(f_path, session_id, parts_header, gql)))
+	
+}
+
+get_gql_header <- function() {
+	httr::add_headers(
+		.headers = c(`content-type` = "application/json", 
+								 accept = "application/json", 
+								 authorization = paste("Bearer", 
+								 											get_access_token())))
+}
+
+run_gql_query <- function(q, gql) {
+	out <- httr::POST(pkg.env$graphql_url, get_gql_header(), 
+										body = jsonlite::toJSON(list(query = q), auto_unbox = TRUE),  
+										handle = gql)
+	
+	if(out$status_code == 200) {
+		jsonlite::fromJSON(rawToChar(out$content))
+	} else {
+		stop(paste("Error making multipart session.\n code:", out$status_code,
+							 "\n content:", rawToChar(out$content)))		
+	}
+	
+}
+
+#GraphQL Queries for interaction with ScienceBase Manager
+create_multipart_upload_session <- function(s3_filepath, content_type, username, gql) {
+
+	run_gql_query(sprintf(
+		'query { createMultipartUploadSession(object: "%s" contentType: "%s" username: "%s") }', 
+		s3_filepath, content_type, username), gql)$data$createMultipartUploadSession
+
+}
+
+get_presigned_url_for_chunk <- function(s3_filepath, upload_id, part_number, gql) {
+
+	run_gql_query(sprintf(
+		'query { getPreSignedUrlForChunk(object: "%s", upload_id: "%s", part_number: "%s") }',
+		s3_filepath, upload_id, part_number), gql)$data$getPreSignedUrlForChunk
+	
+}
+
+complete_multipart_upload <- function(item_str, upload_id, etag_payload, gql) {
+	
+	eta <- sapply(etag_payload, function(x) {
+		sprintf('{ETag: "%s", PartNumber: %i}', gsub('"', "", x$ETag), x$PartNumber)
+	})
+	
+	eta <- paste0("[", paste(eta, collapse = ","), "]")
+	
+	run_gql_query(sprintf(
+		'query { completeMultiPartUpload(object: "%s" upload_id: "%s" parts_eTags: %s) }',
+		item_str, upload_id, eta), gql)
+}
```

Comment:
These gql queries were finicky as hell. Probably a better way to create them.

Reply:
There are some R gql clients out there, e.g. this one: https://github.com/ropensci/ghql.  Maybe they would abstract some of that query construction for you?  I basically do what you're doing here for SQL commands though, it gets the job done.